File: apps/mobile-godot/plugins/ios/MEDdonishPlugin.swift
import Foundation

/// iOS native shell for Godot.
/// AVPlayer + FairPlay, StoreKit, Keychain, APNs — compiled on macOS + Xcode.
/// Must not grant entitlement. Must not log tokens, OTP, or DRM keys.
@objc(MEDdonishPlugin)
public final class MEDdonishPlugin: NSObject {
    @objc public static let pluginName = "MEDdonish"

    @objc public func play_video(_ playbackUrl: String, watermark: String) {
        precondition(!playbackUrl.isEmpty)
        precondition(!watermark.isEmpty)
        // AVPlayer + FairPlay. Disable screen capture analog. playbackUrl is a signed session, not a storage key.
    }

    @objc public func integrity_ticket() -> String {
        // App Attest / DeviceCheck → HMAC ticket for x-app-integrity. Empty until native SDK is wired.
        return ""
    }

    @objc public func start_purchase(_ productId: String, orderId: String) {
        precondition(!productId.isEmpty)
        precondition(!orderId.isEmpty)
        // StoreKit. After the store returns a token, Godot calls MEDdonishHost.submit_store_token.
        // Access only after App Store Server Notification → payment.status = paid.
    }

    @objc public func secure_put(_ key: String, value: String) {
        precondition(!key.isEmpty)
        precondition(!value.isEmpty)
        // Keychain. Refresh token must not be written to a file.
    }

    @objc public func secure_get(_ key: String) -> String {
        precondition(!key.isEmpty)
        return ""
    }

    @objc public func register_push() {
        // APNs. When a token is available, Godot calls MEDdonishHost.submit_push_token. Never log it.
    }
}

File: apps/mobile-godot/plugins/android/MEDdonishPlugin.kt
package com.meddonish.godot

import org.godotengine.godot.Godot
import org.godotengine.godot.plugin.GodotPlugin
import org.godotengine.godot.plugin.SignalInfo
import org.godotengine.godot.plugin.UsedByGodot

/**
 * Android native shell for Godot.
 * Media3 + Widevine, Play Billing, Android Keystore, FCM — wired on a machine with Android SDK.
 * This class must not grant entitlement and must not log tokens, OTP, or DRM keys.
 */
class MEDdonishPlugin(godot: Godot) : GodotPlugin(godot) {
    override fun getPluginName() = "MEDdonish"

    override fun getPluginSignals(): Set<SignalInfo> = setOf(
        SignalInfo("purchase_token", String::class.java, String::class.java, String::class.java),
        SignalInfo("push_token", String::class.java),
    )

    @UsedByGodot
    fun play_video(playbackUrl: String, watermark: String) {
        require(playbackUrl.isNotBlank())
        require(watermark.isNotBlank())
        // Media3 / ExoPlayer + Widevine + FLAG_SECURE. playbackUrl is a signed session, not a storage key.
    }

    @UsedByGodot
    fun integrity_ticket(): String {
        // Play Integrity → HMAC ticket for x-app-integrity. Empty until native SDK is wired.
        return ""
    }

    @UsedByGodot
    fun start_purchase(sku: String, orderId: String) {
        require(sku.isNotBlank())
        require(orderId.isNotBlank())
        // Play Billing. After the store returns a token, emit purchase_token(orderId, GOOGLE_PLAY, token).
        // Godot posts /orders/:id/store-receipt. Access only after webhook paid.
    }

    @UsedByGodot
    fun secure_put(key: String, value: String) {
        require(key.isNotBlank())
        require(value.isNotBlank())
        // Android Keystore. Refresh token must not go to SharedPreferences plaintext.
    }

    @UsedByGodot
    fun secure_get(key: String): String {
        require(key.isNotBlank())
        return ""
    }

    @UsedByGodot
    fun register_push() {
        // FCM. When a token is available emit push_token(token). Never log it.
    }

    fun emitStoreToken(orderId: String, token: String) {
        emitSignal("purchase_token", orderId, "GOOGLE_PLAY", token)
    }

    fun emitPushToken(token: String) {
        emitSignal("push_token", token)
    }
}
